### DISCUSSION TITLE
C# Coding Style - fields (which naming convention)?

### DISCUSSION THREAD

**Discussion by DrGriff** [dotnet/runtime#54140] [Q&A] [answered]
In [C# Coding Style](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md) it states:

> 3. We use _camelCase for internal and private fields and use readonly where possible. Prefix internal and private instance fields with `_`, static fields with `s_` and thread static fields with `t_`. 

However, from Microsoft's Naming Guidelines for [Names of Fields](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members#names-of-fields) it states:

> DO NOT use a prefix for field names.
For example, do not use "g_" or "s_" to indicate static fields.

Err, so, these contradict each other.  Surely Microsoft's guidelines for C# usage are the ones to follow....but both these docs are authored by Microsoft, right?

**Comment by stephentoub** [comment, marked as answer]
See the first sentence of the Names of Fields section:
"The field-naming guidelines apply to static public and protected fields. Internal and private fields are not covered by guidelines"

And the cited C# Coding Style phrase you highlighted:
"for internal and private fields"

They're not in conflict.  The former is about things exposed externally from an assembly, whereas the latter is about implementation detail.

**Reply by FlashGordon67** [reply]
I would likt to add, that if you are looking for C# Coding Conventions, you should not only refer to the .NET Framework Guidelines, but also to the C# documentation from Microsoft. 

Here are some links:
https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/identifier-names
https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions

If you open the latter link you will find that Microsoft has just recently (May 2021) updated their coding conventions and now explicitly states the same rules for internal and private field names as in the dotnet/runtime C# Coding Style guide. Microsoft even cites it on top of the page as the source it referred to.

But I would also like to add another issue in referral of No 3 [C# Coding Style ](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md): From my point of view, the stated conventions do not address protected static fields. As to Microsoft's Framework Guidelines these are allowed and should use PascalCasing. I guess, when stating "...static fields with s_ and thread static fields with t_" you mean private static and internal static fields, only. So, to me, the naming of protected static fields, is missing.

**Reply by stephentoub** [reply]
When the dotnet/runtime guidelines refer to style or compatibility around public surface area, it's really talking about anything exposed for external consumption... for C# visibility, that includes "public", "protected", and "protected internal".
